packages/core/src/modules/connections/services/DidRotateService.ts

```diff
@@ -0,0 +1,265 @@
+import type { Routing } from './ConnectionService'
+import type { AgentContext } from '../../../agent'
+import type { InboundMessageContext } from '../../../agent/models/InboundMessageContext'
+import type { ConnectionRecord } from '../repository/ConnectionRecord'
+
+import { EventEmitter } from '../../../agent/EventEmitter'
+import { OutboundMessageContext } from '../../../agent/models'
+import { InjectionSymbols } from '../../../constants'
+import { AriesFrameworkError } from '../../../error'
+import { Logger } from '../../../logger'
+import { inject, injectable } from '../../../plugins'
+import { AckStatus } from '../../common'
+import {
+  DidRepository,
+  DidResolverService,
+  PeerDidNumAlgo,
+  getAlternativeDidsForPeerDid,
+  getNumAlgoFromPeerDid,
+  isValidPeerDid,
+} from '../../dids'
+import { getMediationRecordForDidDocument } from '../../routing/services/helpers'
+import { ConnectionsModuleConfig } from '../ConnectionsModuleConfig'
+import { RotateMessage, RotateAckMessage, DidRotateProblemReportMessage, HangupMessage } from '../messages'
+import { ConnectionMetadataKeys } from '../repository/ConnectionMetadataTypes'
+
+import { ConnectionService } from './ConnectionService'
+import { createPeerDidFromServices, getDidDocumentForCreatedDid, routingToServices } from './helpers'
+
+@injectable()
+export class DidRotateService {
+  private eventEmitter: EventEmitter
+  private didResolverService: DidResolverService
+  private logger: Logger
+
+  public constructor(
+    eventEmitter: EventEmitter,
+    didResolverService: DidResolverService,
+    @inject(InjectionSymbols.Logger) logger: Logger
+  ) {
+    this.eventEmitter = eventEmitter
+    this.didResolverService = didResolverService
+    this.logger = logger
+  }
+
+  public async createRotate(
+    agentContext: AgentContext,
+    options: { connection: ConnectionRecord; did?: string; routing: Routing }
+  ) {
+    const { connection, did, routing } = options
+
+    const config = agentContext.dependencyManager.resolve(ConnectionsModuleConfig)
+
+    // Do not allow to receive concurrent did rotation flows
+    const didRotateMetadata = connection.metadata.get(ConnectionMetadataKeys.DidRotate)
+
+    if (didRotateMetadata) {
+      this.logger.warn(`There is already an existing opened did rotation flow for connection id ${connection.id}`)
```

Comment:
Should this return? As you said do not allow concurrent flows

Reply:
Yes, I think it is better to throw error in that case. I didn't want to be so strict, in order to allow restarting the flow in case the ack is never received (or the rotate message couldn't be delivered), but I'll add a method at `DidRotateService` level to clear rotation data so it will be somehow possible from a client to reset the flow without getting into details about the metadata.